### ISSUE TITLE
RFE: Allow configuration of OCSP client request timeout

### ISSUE THREAD

**Issue by cipherboy** [openbao/openbao#51]
As reported by @celesteking on `https://github.com/hashicorp/vault/pull/17093#issuecomment-1897497036`:

> Can we please get configurable ocsp request timeouts? Imagine ocsp responder being down (firewalled, incorrectly configured, etc), your ocsp client would be banging it 5 times before giving up. And the cert auth would hang for ~1 min. 

(edited for brevity)

So far only cert auth uses this OCSP client, so it would be the only place we'd need to expose this configuration, similar to existing configuration parameters.

**Comment by cipherboy** [comment]
@celesteking Some commentary on this: 

The OCSP package uses [`go-retryablehttp.Client`](https://pkg.go.dev/github.com/hashicorp/go-retryablehttp#Client) as its HTTP client; this has a per-request timeout plus a max retries which defaults to 4. Currently this is set to 10 seconds and 4 retries, but each additional fallback server will add to the latency. With one server, you'd see ~40s of latency, and each additional server will add to that (unless you specify `QueryAllServers`, in which case an error will still be bubbled up instead of an OCSP response).

Ultimately, I think this means that the best approach might come from a more persistent OCSP client implementation. For one, we'd need a stateful client to track what the "best" OCSP server is (e.g., if one is consistently reporting errors, switch to another by default). Or,  we could swap out our retryable HTTP client to interleave requests to multiple services (e.g., send req 1 to server 1, then req 1 to server 2, before coming back and attempting retries on either). 

The latter isn't very common scenario, IMO, so I'm not sure if upstream would accept it as a change to the library... But we can certainly try.

On the other hand, I _can_ expose the retry and timeout configuration to the cert auth method, but based on the above, I don't think it'd quite be optimal: if server 1 & 2 are down but 3 is OK, for one, that'd still report as an error (in the parallel query method), and for two, we'd still have to sit through `retries * timeout * down server` attempts prior to finally contacting a correct server.

Thoughts?

**Comment by celesteking** [comment]
Appreciate the thorough review. I'm an end-user and have little idea how to approach this, but what I can tell for sure is that there are people out there eager to exploit this feature. Put a hacker hat on and try to DoS the server by making thousands of parallel requests with aia ocsp uri containing an arbitrary host, say 2.2.2.2, that will never reply.

There should be some timeout controls exposed, I believe.

**Comment by celesteking** [comment]
Another (offtopic?) concern I've got is related to lack of coordination between `auth/cert` and `pki` components. As I understand, currently they're independent, which means ocsp on `auth` MUST be on in order to refuse revoked certificates in `pki`. It would be wonderful to "tie" them so that ocsp could be skipped altogether and have a setting to deny "foreign" certificates outright.

**Comment by cipherboy** [comment]
@celesteking Sorry about the delay in getting back to you. :-) 

OpenBao (and upstream) lack cross-plugin communication of any sort; this is well documented on various issues upstream and hard to resolve, but if you want to put forward a very concrete [proposal](https://github.com/openbao/openbao/issues/new?assignees=&labels=rfc%2Cpending-decision&projects=&template=rfc.yml) for doing so, I'd enjoy reading it...

You're partially correct about requiring OCSP to be on to get revocation information -- CRLs are the other, alternative mechanism and are more fault tolerant and what I'd generally recommend.

OCSP works great as a TLS client, where you are connecting to many servers, each of which may be signed by a different arbitrary CA; there's lots of CAs on the internet and a vast number of servers, and so it hardly makes sense to download a potentially many gigabyte+ CRL for each CA (ok, I'm exaggerating a little here!). For every server you connect to this (potentially, OCSP stapling avoids it) incurs _another_ network round trip, but that's OK in the grand scheme of HTTPS and many-megabyte websites sprawling over many domains with CDNs and the like.

On the other hand, CRLs work great as a TLS server enforcing client auth, wherein you know you're getting connections from one of a few handful of trusted CAs (and rejecting the rest). Only for these trusted CAs would you maintain CRLs (and not automatically fetch them -- and the TLS auth method supports reloading them via configured URLs. And from the few CAs you have, you're likely to get many client connections, so keeping the whole CRL locally (which should only be a small, small fraction of connections -- revocations should be rare and lifetimes short to avoid them) is much more beneficial. Think of this more like a hashmap lookup.

An online round trip is thus much less appealing when the performance could be as fast as a hashmap lookup, hence, I'd suggest using CRLs for this case, and giving OpenBao the URL to auto-update the CRLs. :-)

(As an aside, there's a few cases where you _really_ want that up-to-the-second OCSP lookup, but even OCSP isn't necessarily guaranteed to be instantaneous and can be cached by either the client or pushed from the CA out to a caching CDN... and in that case, you really want to consider failure modes, ensuring fail-closed, and overall system uptime which is hard to do in a project without replication :-).

So the TL;DR: I'd use CRLs and they should avoid most of this pain.

**Comment by cipherboy** [comment]
> Appreciate the thorough review. I'm an end-user and have little idea how to approach this, but what I can tell for sure is that there are people out there eager to exploit this feature. Put a hacker hat on and try to DoS the server by making thousands of parallel requests with aia ocsp uri containing an arbitrary host, say 2.2.2.2, that will never reply.
> 
> There should be some timeout controls exposed, I believe.

I'd be curious in a reproducer here if you have one.

If my memory of this auth plugin is correct, this would _only_ be a valid attack if you have an outright malicious CA here, _that you trust_. It is essentially a self-own. Leaves from untrusted CAs won't have their AIAs pulled as they are already rejected at chainbuilding time, prior to any OCSP lookup (and we do not support using AIAs to discover missing ICAs/...). This leaves leaves from trusted-but-acting-maliciously-CAs, which... seems like a conversation you should really have with the people that run the CA. :-)

IOW, I'm happy to be proven wrong, but I don't think this is actually a valid attack.

**Comment by celesteking** [comment]
You're right regarding this being more of a self-inflicted issue rather than a vulnerability. Indeed, AIA (ocsp in my case) check happens post chain-building and you can't just trigger it with a random certificate from a random CA.

In my case, we were just switching from old pki system to Vault's PKI and thus old clients using previously issued certs had the old CA's AIA. Vault hosts have been firewalled off not allowing any external communication, thus the timeouts and thus the issue.

I appreciate the time spent on clearing things up for me.

> lack cross-plugin communication of any sort

BTW, maybe that's for good, at least in terms of security. Zero trust paradigm has proven to be a successful one.

**Comment by cipherboy** [comment]
> BTW, maybe that's for good, at least in terms of security. Zero trust paradigm has proven to be a successful one.

This is true, but there is a gradient, where perhaps the existing token is used to re-authenticate to other endpoints, perhaps driven by this plugin. Think of it more as server-driven request chaining. Here, ACL system could still apply (perhaps with some new designation for whether another plugin could see the response of the request).

---

That said, I think it could be worthwhile to build a better OCSP client. I might have to think about it a bit more.
